src/Database/LSMTree/Internal/Serialise/RawBytes.hs

```diff
@@ -0,0 +1,193 @@
+{-# LANGUAGE BangPatterns               #-}
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs               #-}
+{-# LANGUAGE MagicHash                  #-}
+{-# LANGUAGE UnboxedTuples              #-}
+{- HLINT ignore "Redundant lambda" -}
+
+#include <MachDeps.h>
+
+module Database.LSMTree.Internal.Serialise.RawBytes (
+    RawBytes (..)
+  , topBits16
+  , sliceBits32
+  , sizeofRawBytes
+  , Database.LSMTree.Internal.Serialise.RawBytes.concat
+    -- * @bytestring@ utils
+  , unsafeFromByteString
+  , fromShortByteString
+  , rawBytes
+  ) where
+
+import           Control.DeepSeq
+import           Control.Exception (assert)
+import           Data.Bits (Bits (shiftL, shiftR))
+import           Data.BloomFilter.Hash (hashList32)
+import           Data.ByteString as BS
+import qualified Data.ByteString.Builder as BB
+import qualified Data.ByteString.Builder.Internal as BB
+import           Data.ByteString.Internal as BS.Internal
+import           Data.ByteString.Short (ShortByteString (SBS))
+import qualified Data.ByteString.Short as SBS
+import qualified Data.ByteString.Short.Internal as SBS
+import           Data.Primitive.ByteArray (ByteArray (..), compareByteArrays)
+import qualified Data.Vector.Primitive as P
+import           Database.LSMTree.Internal.Run.BloomFilter (Hashable (..))
+import           Foreign.Ptr
+import           GHC.Exts
+import           GHC.ForeignPtr as GHC
+import           GHC.Word
+
+newtype RawBytes = RawBytes (P.Vector Word8)
```

Comment:
```suggestion
-- |Raw bytes with no alignment constraint (i.e. byte aligned), and no guarantee
-- of pinned or unpinned memory (i.e. could be either).
--
newtype RawBytes = RawBytes (P.Vector Word8)
```
Since this is going to be the user-visible type for serialisation, the user will need to know this (if they're doing low level high performance serialisation tricks).